src/core/shutdown.c

```diff
@@ -159,6 +163,78 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_free_ char *dirty_string = NULL;
+        unsigned long long dirty = 0;
+        int r;
+
+        r = get_proc_field("/proc/meminfo", "Dirty", WHITESPACE, &dirty_string);
+        if (r < 0) {
+                log_error_errno(r, "Failed to retrieve Dirty from /proc/meminfo: %m");
+                return false;
+        }
+
+        r = safe_atollu(dirty_string, &dirty);
+        if (r < 0) {
+                log_error_errno(r, "Failed to parse Dirty from /proc/meminfo: %s: %m",
+                                dirty_string);
+                return false;
+        }
+
+        if (*prev_dirty > dirty)
+                r = true;
+        else
+                r = false;
+
+        *prev_dirty = dirty;
+
+        return r;
+}
+
+static void sync_with_progress(void) {
+        unsigned checks;
+        pid_t pid;
+        int r;
+        unsigned long long dirty = -1;
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        /* Due to the possiblity of the sync operation hanging, we fork
+         * a child process and monitor the progress. If the timeout
+         * lapses, the assumption is that that particular sync stalled. */
+        pid = fork();
+        if (pid < 0)
+                log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                /* Start the sync operation here in the child */
+                log_info("Syncing filesystems and block devices");
+                sync();
+                _exit(EXIT_SUCCESS);
+        }
+
+        /* Start monitoring the sync operation. If more than
+         * SYNC_PROGRESS_ATTEMPTS lapse without the /proc/meminfo "Dirty"
+         * value decreasing, we assume that the sync is stalled */
+        for (checks = 0; checks < SYNC_PROGRESS_ATTEMPTS; checks++) {
+                r = wait_for_terminate_with_timeout(pid, SYNC_TIMEOUT_USEC);
+                if (r < 0 && r == -ETIMEDOUT) {
+                        /* Reset the check counter if the "Dirty" value is
+                         * decreasing */
+                        if (sync_making_progress(&dirty) == true)
```

Comment:
C's `if` concept already checks if things are true, hence, just drop the ` == true` it's entirely redundant...